satpy/readers/iasi_l2_so2_bufr.py

```diff
@@ -0,0 +1,235 @@
+        """Gets the first and last date from the bufr file."""
+        fh = open(self.filename, "rb")
+        i = 0
+        while True:
+            # get handle for message
+            bufr = ec.codes_bufr_new_from_file(fh)
+            if bufr is None:
+                break
+            ec.codes_set(bufr, 'unpack', 1)
+            year = ec.codes_get(bufr, 'year')
+            month = ec.codes_get(bufr, 'month')
+            day = ec.codes_get(bufr, 'day')
+            hour = ec.codes_get(bufr, 'hour')
+            minute = ec.codes_get(bufr, 'minute')
+            second = ec.codes_get(bufr, 'second')
+
+            obs_time = datetime(year=year, month=month, day=day, hour=hour, minute=minute, second=second)
+
+            if i == 0:
+                start_time = obs_time
+
+            ec.codes_release(bufr)
+
+            i += 1
+
+        end_time = obs_time
+
+        fh.close()
+
+        return start_time, end_time
+
+    def get_attribute(self, key):
+        """Get BUFR attributes."""
+        # This function is inefficient as it is looping through the entire
+        # file to get 1 attribute. It causes a problem though if you break
+        # from the file early - dont know why but investigating - fix later
+
+        fh = open(self.filename, "rb")
+        while True:
+            # get handle for message
+            bufr = ec.codes_bufr_new_from_file(fh)
+            if bufr is None:
+                break
+            ec.codes_set(bufr, 'unpack', 1)
+            attr = ec.codes_get(bufr, key)
+            ec.codes_release(bufr)
+
+        fh.close()
+        return attr
+
+    def get_array(self, key):
+        """Get all data from file for the given BUFR key."""
+        with open(self.filename, "rb") as fh:
+            msgCount = 0
+            while True:
+
+                bufr = ec.codes_bufr_new_from_file(fh)
+                if bufr is None:
+                    break
+
+                ec.codes_set(bufr, 'unpack', 1)
+
+                values = ec.codes_get_array(
+                        bufr, key, float)
+
+                if len(values) == 1:
+                    values = np.repeat(values, 120)
+
+                # if is the first message initialise our final array
+                if (msgCount == 0):
+
+                    arr = da.from_array([values], chunks=CHUNK_SIZE)
+                else:
+                    tmpArr = da.from_array([values], chunks=CHUNK_SIZE)
+
+                    arr = da.concatenate((arr, tmpArr), axis=0)
+
+                msgCount = msgCount+1
+                ec.codes_release(bufr)
+
+        if arr.size == 1:
+            arr = arr[0]
+
+        return arr
+
+    def get_dataset(self, dataset_id, dataset_info):
+        """Get dataset using the BUFR key in dataset_info."""
+        arr = self.get_array(dataset_info['key'])
+        arr[arr == dataset_info['fill_value']] = np.nan
+
+        xarr = xr.DataArray(arr, dims=["y", "x"], name=dataset_info['name'])
```

Comment:
We like to work with dask array in satpy, so make sure `arr` is one before you embed it in a `DataArray`

Reply:
Thanks Martin, the function `get_array` returns a dask array, and adding

`arr[arr == dataset_info['fill_value']] = np.nan`

at line 228 seems to also produce a dask array, but I'm wondering if there's some way it could be being converted back to a numpy array that I'm missing? 

Is the idea to check the type and and throw an exception if it's not a dask array, say at line 229?

Reply:
No, if `get_array` returns a dask array, is should be fine. But I'm puzzled on why line 228 isn't crashing, I didn't think dask arrays could do in-place assignments.

Reply:
ah, that's the issue, I'll look into it ...

Reply:
It seems it depends how you do it, if you run this:

```
import numpy as np
import dask.array as da
x = np.array([1,2,3,4,5])
CHUNK_SIZE=2
y = da.from_array(x, chunks=CHUNK_SIZE)
```
then 
`y[2] = -99`
throws an error, but this does not:
```
y[y==2.] = -99.
```

`y.compute() `gives what you would expect. Do you think it saves the replacement and implements it when you compute?

Reply:
Seems it's working as it should then, thanks for testing it!